### PR TITLE
Add initial version of cyclonedx format

### PR DESCRIPTION
**Pull request by mateuszdyminski**
Changes:

- Added custom `ksoc:kbom:` taxonomy as extension of https://github.com/CycloneDX/cyclonedx-property-taxonomy
- Added 2 new formats: `cyconedx-json` and `cyclonedx-xml`
- Added new flag `context` which allows to fetch `kbom` file for particular k8s context

Towards:

- https://github.com/ksoc-private/product-issues/issues/37

### CODE REVIEW

File: cmd/cyclonexdx.go [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+				{
+					Name:  KSOCPrefix + "k8s:node:containerruntimeversion",
+					Value: n.ContainerRuntimeVersion,
+				},
+				{
+					Name:  KSOCPrefix + "k8s:node:kubeletversion",
+					Value: n.KubeletVersion,
+				},
+				{
+					Name:  KSOCPrefix + "k8s:node:kuebproxyversion",
```
Comment (jimmed) [inline cmd/cyclonexdx.go]: This looks like a typo
```suggestion
					Name:  KSOCPrefix + "k8s:node:kubeproxyversion",
```

File: docs/taxonomy.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+| `ksoc:kbom:k8s:node:arch`                      | Node's architecture               |
+| `ksoc:kbom:k8s:node:kernel`                    | Node's kernel version             |
+| `ksoc:kbom:k8s:node:bootid`                    | Node's Boot identifier            |
+| `ksoc:kbom:k8s:node:type`                      | Node's type                       |
+| `ksoc:kbom:k8s:node:operatingsystem`           | Node's operating system           |
+| `ksoc:kbom:k8s:node:machineid`                 | Node's machine identifier         |
+| `ksoc:kbom:k8s:node:hostname`                  | Node's hostaname                  |
+| `ksoc:kbom:k8s:node:containerruntimeversion`   | Node's container runtime version  |
+| `ksoc:kbom:k8s:node:kubeletversion`            | Node's kubelet version            |
+| `ksoc:kbom:k8s:node:kuebproxyversion`          | Node's kubelet version            |
```
Comment (jimmed) [inline docs/taxonomy.md]:
```suggestion
| `ksoc:kbom:k8s:node:kubeproxyversion`          | Node's kubelet version            |
```

Comment (jimmed) [reply docs/taxonomy.md]: The description looks to be wrong here too

File: cmd/cyclonexdx.go (hunk 1 above)
Comment (mateuszdyminski) [reply cmd/cyclonexdx.go]: fixed

File: docs/taxonomy.md (hunk 2 above)
Comment (mateuszdyminski) [reply docs/taxonomy.md]: fixed

File: docs/taxonomy.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+| Property                                       | Description                       |
+| ---------------------------------------------- | --------------------------------- |
+| `ksoc:kbom:k8s:node:osimage`                   | Node's operating system image     |
+| `ksoc:kbom:k8s:node:arch`                      | Node's architecture               |
+| `ksoc:kbom:k8s:node:kernel`                    | Node's kernel version             |
+| `ksoc:kbom:k8s:node:bootid`                    | Node's Boot identifier            |
+| `ksoc:kbom:k8s:node:type`                      | Node's type                       |
+| `ksoc:kbom:k8s:node:operatingsystem`           | Node's operating system           |
+| `ksoc:kbom:k8s:node:machineid`                 | Node's machine identifier         |
+| `ksoc:kbom:k8s:node:hostname`                  | Node's hostaname                  |
```
Comment (jimmed) [inline docs/taxonomy.md]:
```suggestion
| `ksoc:kbom:k8s:node:hostname`                  | Node's hostname                  |
```

File: docs/taxonomy.md [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+| `ksoc:kbom:k8s:node:containerruntimeversion`   | Node's container runtime version  |
+| `ksoc:kbom:k8s:node:kubeletversion`            | Node's kubelet version            |
+| `ksoc:kbom:k8s:node:kubeproxyversion`          | Node's kubelet version            |
```
Comment (jimmed) [inline docs/taxonomy.md]:
```suggestion
| `ksoc:kbom:k8s:node:kubeproxyversion`          | Node's kubeproxy version          |
```

Comment (mateuszdyminski) [reply docs/taxonomy.md]: fixed

File: docs/taxonomy.md (hunk 3 above)
Comment (mateuszdyminski) [reply docs/taxonomy.md]: fixed

File: internal/kube/kube.go [hunk 5]
```diff
@@ -116,7 +142,7 @@ func (k *k8sDB) AllNodes(ctx context.Context, full bool) ([]model.Node, error) {
 			MachineID:               nodes.Items[i].Status.NodeInfo.MachineID,
 			Architecture:            nodes.Items[i].Status.NodeInfo.Architecture,
 			KernelVersion:           nodes.Items[i].Status.NodeInfo.KernelVersion,
-			ContainerRuntimeVersion: nodes.Items[i].Status.NodeInfo.ContainerRuntimeVersion,
+			ContainerRuntimeVersion: strings.TrimPrefix(nodes.Items[i].Status.NodeInfo.ContainerRuntimeVersion, "containerd://"),
```
Comment (jeffreyfriedman) [inline internal/kube/kube.go]: Do we need to handle the other possible container runtimes here?

File: internal/kube/kube.go [hunk 6]
```diff
@@ -206,13 +234,34 @@ func containerToImage(img, imgName string, statuses []v1.ContainerStatus) (*mode
 			if strings.Contains(statuses[i].ImageID, "@") {
 				res.Digest = strings.Split(statuses[i].ImageID, "@")[1]
 			}
+			if strings.HasPrefix(statuses[i].ImageID, "sha256:") {
+				res.Digest = statuses[i].ImageID
+			}
```
Comment (jeffreyfriedman) [inline internal/kube/kube.go]: Are these scenarios mutually exclusive, or should we `break` after L235?

Comment (mateuszdyminski) [reply internal/kube/kube.go]: they are exclusive - but I have added `else` to make it easier to read

File: internal/kube/kube.go (hunk 5 above)
Comment (mateuszdyminski) [reply internal/kube/kube.go]: removed that prefix trimming - with it we loose information about the runtime at all 

in the future we should add dependency to containerd or other runtime with full description

File: cmd/schema_test.go [hunk 7]
```diff
@@ -81,34 +84,66 @@ var expectedSchema = `{
...
+            "$ref": "#/$defs/Image"
+          },
+          "type": "array"
+        },
+        "resources": {
+          "patternProperties": {
+            ".*": {
+              "$ref": "#/$defs/ResourceList"
+            }
+          },
```
Comment (jimmed) [inline cmd/schema_test.go:116]: The API client generator isn't going to understand this (yet); the type is going to resolve as an empty object.

**Review by jimmed — APPROVED** [review]

File: cmd/schema_test.go (hunk 7 above)
Comment (mateuszdyminski) [reply cmd/schema_test.go:116]: Not sure if we will use this schema as input for out generators?

Comment (mateuszdyminski) [reply cmd/schema_test.go:116]: as it's used only as `kbom schema` output - closing discussion

**Pull request merged**